### ISSUE TITLE
Cannot read property '0' of undefined

### ISSUE THREAD

**Issue by rvdblom** [AdyRock/com.onvif#9]
Hello,

I am unable to add an onvif camera

I have made a rapport

f7943ec5-7ca3-429a-aab6-2a2dca0706cc

Log ID: 2020-09-12T07:29:45.193Z
App version 1.1.1

+0:0:0.000: Discovery error: Wrong SOAP message from 10.0.1.26:3702
+0:0:14.489: Connection Failed for 10.0.1.26 Port: 8000 Username: Admin
+0:0:0.017: Failed: TypeError: Cannot read property '0' of undefined
    at Cam.<anonymous> (/lib/onvif/lib/cam.js:373:104)
    at /lib/onvif/lib/utils.js:103:6
    at Parser.<anonymous> (/node_modules/xml2js/lib/parser.js:304:18)
    at Parser.emit (events.js:311:20)
    at SAXParser.onclosetag (/node_modules/xml2js/lib/parser.js:262:26)
    at emit (/node_modules/sax/lib/sax.js:624:35)
    at emitNode (/node_modules/sax/lib/sax.js:629:5)
    at closeTag (/node_modules/sax/lib/sax.js:889:7)
    at SAXParser.write (/node_modules/sax/lib/sax.js:1436:13)
    at Parser.exports.Parser.Parser.parseString (/node_modules/xml2js/lib/parser.js:323:31)
    at Parser.parseString (/node_modules/xml2js/lib/parser.js:5:59)
    at Object.exports.parseString (/node_modules/xml2js/lib/parser.js:369:19)
    at parseSOAPString (/lib/onvif/lib/utils.js:59:10)
    at IncomingMessage.<anonymous> (/lib/onvif/lib/cam.js:301:5)
    at IncomingMessage.emit (events.js:323:22)
    at endReadableNT (_stream_readable.js:1204:12)
    at processTicksAndRejections (internal/process/task_queues.js:84:21)

Could you look in to this?

If you need more information please let me know

Thanks in advance

**Comment by rvdblom** [comment]
Website: https://www.orretti.com/product/orretti-wifi-camera-hd-2mp-met-ios-android-smart-app-draadloze-ip-video-beveiligingscamera-met-nachtzicht-bewegingsdetectie-babyfoon-wit/

**Comment by AdyRock** [comment]
Thanks for following up on the crash report. I will look into this today and get back to you if I have any questions.

**Comment by AdyRock** [comment]
Hi, Could you make sure logging is enabled in the app configuration page, try to add the camera and then send the log to me.
Hopefully that will show me what is coming back from the camera so I can see why it is failing to parse the message.

**Comment by rvdblom** [comment]
Thanks for your fast response!

Did you mean this log?

120: ====  Discovery Starting  ====
+0:0:0.133: Reply from {
  "hostname": "10.0.1.26",
  "port": "8000",
  "path": "/onvif/device_service",
  "urn": "urn:uuid:06025677-2dd2-21b2-a205-4A1599906466",
  "href": "http://10.0.1.26:8000/onvif/device_service"
}
+0:0:0.017: Discovery error: Wrong SOAP message from 10.0.1.26:3702
+0:0:4.879: ====  Discovery Finished  ====
+0:0:0.002: Discovered: [
  {
    "name": "10.0.1.26",
    "data": {
      "id": "10.0.1.26",
      "port": "8000"
    },
    "settings": {
      "username": "",
      "password": "",
      "ip": "10.0.1.26",
      "port": "8000",
      "urn": "urn:uuid:06025677-2dd2-21b2-a205-4A1599906466",
      "channel": 1
    }
  }
]
+0:0:9.074: Login-----
+0:0:0.015: --------------------------
+0:0:0.010: Connect to Camera 10.0.1.26:8000 - admin
+0:0:0.058: Connection Failed for 10.0.1.26 Port: 8000 Username: admin
+0:0:0.013: Failed: TypeError: Cannot read property '0' of undefined
    at Cam.<anonymous> (/lib/onvif/lib/cam.js:373:104)
    at /lib/onvif/lib/utils.js:103:6
    at Parser.<anonymous> (/node_modules/xml2js/lib/parser.js:304:18)
    at Parser.emit (events.js:311:20)
    at SAXParser.onclosetag (/node_modules/xml2js/lib/parser.js:262:26)
    at emit (/node_modules/sax/lib/sax.js:624:35)
    at emitNode (/node_modules/sax/lib/sax.js:629:5)
    at closeTag (/node_modules/sax/lib/sax.js:889:7)
    at SAXParser.write (/node_modules/sax/lib/sax.js:1436:13)
    at Parser.exports.Parser.Parser.parseString (/node_modules/xml2js/lib/parser.js:323:31)
    at Parser.parseString (/node_modules/xml2js/lib/parser.js:5:59)
    at Object.exports.parseString (/node_modules/xml2js/lib/parser.js:369:19)
    at parseSOAPString (/lib/onvif/lib/utils.js:59:10)
    at IncomingMessage.<anonymous> (/lib/onvif/lib/cam.js:301:5)
    at IncomingMessage.emit (events.js:323:22)
    at endReadableNT (_stream_readable.js:1204:12)
    at processTicksAndRejections (internal/process/task_queues.js:84:21)
+0:0:21.235: Sending log
+0:0:3.727: Message sent: <2a1b671a-f19c-c063-e369-84e0f811f62b@ady.duckdns.org>


I also push the button “send log”

**Comment by AdyRock** [comment]
That is the correct log, thanks. Unfortunately it's still not showing the source of the error so I have published a new test version with more logging in it https://homey.app/a/com.onvif/test/
Could you install that and try again. Hopefully that will show more detail around the area it is failing.
Thanks.

**Comment by rvdblom** [comment]
I have installed the app and got this log

Log ID: 2020-09-12T11:14:04.309Z
App version 1.1.2

+0:0:0.000: ====  Discovery Starting  ====
+0:0:0.148: Reply from {
  "hostname": "10.0.1.26",
  "port": "8000",
  "path": "/onvif/device_service",
  "urn": "urn:uuid:06025677-2dd2-21b2-a205-4A1599909249",
  "href": "http://10.0.1.26:8000/onvif/device_service"
}
+0:0:4.924: ====  Discovery Finished  ====
+0:0:0.017: Discovered: [
  {
    "name": "10.0.1.26",
    "data": {
      "id": "10.0.1.26",
      "port": "8000"
    },
    "settings": {
      "username": "",
      "password": "",
      "ip": "10.0.1.26",
      "port": "8000",
      "urn": "urn:uuid:06025677-2dd2-21b2-a205-4A1599909249",
      "channel": 1
    }
  }
]
+0:0:9.732: Login-----
+0:0:0.012: --------------------------
+0:0:0.012: Connect to Camera 10.0.1.26:8000 - admin
+0:0:0.081: 
data:
[
  {
    "getSystemDateAndTimeResponse": [
      {
        "systemDateAndTime": [
          ""
        ]
      }
    ]
  }
]
+0:0:0.020: Connection Failed for 10.0.1.26 Port: 8000 Username: admin
+0:0:0.022: Failed: TypeError: Cannot read property '0' of undefined
    at Cam.<anonymous> (/lib/onvif/lib/cam.js:373:104)
    at /lib/onvif/lib/utils.js:103:6
    at Parser.<anonymous> (/node_modules/xml2js/lib/parser.js:304:18)
    at Parser.emit (events.js:311:20)
    at SAXParser.onclosetag (/node_modules/xml2js/lib/parser.js:262:26)
    at emit (/node_modules/sax/lib/sax.js:624:35)
    at emitNode (/node_modules/sax/lib/sax.js:629:5)
    at closeTag (/node_modules/sax/lib/sax.js:889:7)
    at SAXParser.write (/node_modules/sax/lib/sax.js:1436:13)
    at Parser.exports.Parser.Parser.parseString (/node_modules/xml2js/lib/parser.js:323:31)
    at Parser.parseString (/node_modules/xml2js/lib/parser.js:5:59)
    at Object.exports.parseString (/node_modules/xml2js/lib/parser.js:369:19)
    at parseSOAPString (/lib/onvif/lib/utils.js:59:10)
    at IncomingMessage.<anonymous> (/lib/onvif/lib/cam.js:301:5)
    at IncomingMessage.emit (events.js:323:22)
    at endReadableNT (_stream_readable.js:1204:12)
    at processTicksAndRejections (internal/process/task_queues.js:84:21)

**Comment by AdyRock** [comment]
Strange that I not seeing more in the log but it has shown me what part of the get date response is missing.
I have published another test version (same URL) that validates each section of the structure and tries another method instead of failing.
I'm not sure if this will fix the issue or just move it further down the chain but if you can give it a try and send me another log.

**Comment by rvdblom** [comment]
I was able to add the camera now, but got an other errror cannot read property raw of undefined 

Log ID: 2020-09-12T11:54:08.508Z
App version 1.1.3

+0:0:0.000: ====  Discovery Starting  ====
+0:0:0.161: Reply from {
  "hostname": "10.0.1.26",
  "port": "8000",
  "path": "/onvif/device_service",
  "urn": "urn:uuid:06025677-2dd2-21b2-a205-4A1599911654",
  "href": "http://10.0.1.26:8000/onvif/device_service"
}
+0:0:4.879: ====  Discovery Finished  ====
+0:0:0.015: Discovered: [
  {
    "name": "10.0.1.26",
    "data": {
      "id": "10.0.1.26",
      "port": "8000"
    },
    "settings": {
      "username": "",
      "password": "",
      "ip": "10.0.1.26",
      "port": "8000",
      "urn": "urn:uuid:06025677-2dd2-21b2-a205-4A1599911654",
      "channel": 1
    }
  }
]
+0:0:14.458: Login-----
+0:0:0.062: --------------------------
+0:0:0.009: Connect to Camera 10.0.1.26:8000 - admin
+0:0:0.287: getSystemDateAndTime: Finished using Homey time-----
+0:0:0.016: getServices: Started *****
+0:0:0.170: getServices: Finished -----
+0:0:0.019: callUpstartFunctions: Started *****
+0:0:0.386: VideoSources: [
  {
    "$": {
      "token": "VideoSourceConfigSourceToken0"
    },
    "framerate": 25,
    "resolution": {
      "width": 1920,
      "height": 1080
    }
  }
]
+0:0:0.022: ActiveSource [0] = {
  "$": {
    "token": "VideoEncoderConfigToken0"
  },
  "name": "VideoEncoderConfig0",
  "useCount": 1,
  "encoding": "H264",
  "resolution": {
    "width": 1920,
    "height": 1080
  },
  "quality": 50,
  "rateControl": {
    "frameRateLimit": 25,
    "encodingInterval": 1,
    "bitrateLimit": 1200
  },
  "multicast": {
    "address": {
      "type": "IPv4",
      "IPv4Address": "10.0.1.26"
    },
    "port": 8554,
    "TTL": 64,
    "autoStart": true
  },
  "sessionTimeout": "PT0H12M0S"
}
+0:0:0.017: VideoSource [0] = {
  "$": {
    "token": "VideoSourceConfigSourceToken0"
  },
  "framerate": 25,
  "resolution": {
    "width": 1920,
    "height": 1080
  }
}
+0:0:0.015: callUpstartFunctions: Finished -----
+0:0:0.021: CONNECTED to 10.0.1.26
+0:0:0.018: Credentials OK. Adding [
  {
    "$": {
      "token": "VideoSourceConfigSourceToken0"
    },
    "framerate": 25,
    "resolution": {
      "width": 1920,
      "height": 1080
    }
  }
]
+0:0:0.237: list_devices2: Single Sources
+0:0:25.539: Initialising CameraDevice (10.0.1.26)
+0:0:0.060: CameraDevice has been added (10.0.1.26)
+0:0:2.110: Saved Credentials
+0:0:0.043: --------------------------
+0:0:0.019: Connect to Camera 10.0.1.26:8000 - admin
+0:0:0.083: getSystemDateAndTime: Finished using Homey time-----
+0:0:0.013: getServices: Started *****
+0:0:0.076: getServices: Finished -----
+0:0:0.018: callUpstartFunctions: Started *****
+0:0:0.105: VideoSources: [
  {
    "$": {
      "token": "VideoSourceConfigSourceToken0"
    },
    "framerate": 25,
    "resolution": {
      "width": 1920,
      "height": 1080
    }
  }
]
+0:0:0.022: ActiveSource [0] = {
  "$": {
    "token": "VideoEncoderConfigToken0"
  },
  "name": "VideoEncoderConfig0",
  "useCount": 1,
  "encoding": "H264",
  "resolution": {
    "width": 1920,
    "height": 1080
  },
  "quality": 50,
  "rateControl": {
    "frameRateLimit": 25,
    "encodingInterval": 1,
    "bitrateLimit": 1200
  },
  "multicast": {
    "address": {
      "type": "IPv4",
      "IPv4Address": "10.0.1.26"
    },
    "port": 8554,
    "TTL": 64,
    "autoStart": true
  },
  "sessionTimeout": "PT0H12M0S"
}
+0:0:0.018: VideoSource [0] = {
  "$": {
    "token": "VideoSourceConfigSourceToken0"
  },
  "framerate": 25,
  "resolution": {
    "width": 1920,
    "height": 1080
  }
}
+0:0:0.012: callUpstartFunctions: Finished -----
+0:0:0.019: CONNECTED to 10.0.1.26
+0:0:0.151: getServices: Started *****
+0:0:0.102: getServices: Finished -----
+0:0:0.066: Camera Information (10.0.1.26): {
  "manufacturer": "PPS IPC based on ONVIF",
  "model": "Speed 5S",
  "firmwareVersion": "ppstrong-c51-s_neutral-2.9.5.20200205",
  "serialNumber": "060256772",
  "hardwareId": "Ver 2.1"
}
+0:0:0.017: getCapabilities: Started *****
+0:0:0.090: getCapabilities: Finished -----
+0:0:0.019: Camera (10.0.1.26) does NOT support PullPoint Events
+0:0:0.016: Supported Events(10.0.1.26): 
+0:0:0.184: Removing unsupported motion capabilities for 10.0.1.26
+0:0:1.131: Switch motion detection Off (10.0.1.26)
+0:0:0.052: App.unsubscribe: 10.0.1.26
+0:0:0.032: App.unsubscribe: No entry for host: 10.0.1.26
+0:0:0.162: Snapshot URL: http://www.gwelltimes.com/manage/Upload/Pic/20147914514701.jpg
+0:0:0.032: Registering Now image (10.0.1.26)
+0:0:0.032: SnapShot save file (10.0.1.26) = eventimage10_0_1_26.jpg
+0:0:0.020: Initialising event image (10.0.1.26)
+0:0:0.039: Fetching event image (10.0.1.26)
+0:0:0.025: Fetching (10.0.1.26) Motion Event image from: http://www.gwelltimes.com/manage/Upload/Pic/20147914514701.jpg
+0:0:0.059: registered Now image (10.0.1.26)
+0:0:0.953: SnapShot error (10.0.1.26): FetchError: request to http://www.gwelltimes.com/manage/Upload/Pic/20147914514701.jpg failed, reason: socket hang up
    at ClientRequest.<anonymous> (/node_modules/node-fetch/lib/index.js:1455:11)
    at ClientRequest.emit (events.js:311:20)
    at Socket.socketOnEnd (_http_client.js:460:9)
    at Socket.emit (events.js:323:22)
    at endReadableNT (_stream_readable.js:1204:12)
    at processTicksAndRejections (internal/process/task_queues.js:84:21)
+0:0:0.016: Fetch MOTION error (10.0.1.26): {
  "ok": false,
  "statusText": "request to http://www.gwelltimes.com/manage/Upload/Pic/20147914514701.jpg failed, reason: socket hang up"
}
+0:0:0.023: Event SnapShot error (10.0.1.26): Error: request to http://www.gwelltimes.com/manage/Upload/Pic/20147914514701.jpg failed, reason: socket hang up
    at CameraDevice.setupImages (/drivers/camera/device.js:733:13)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at async CameraDevice.connectCamera (/drivers/camera/device.js:290:5)
+0:0:0.029: Camera (10.0.1.26) is ready
+0:0:4.093: getSystemDateAndTime: Finished using Homey time-----
+0:0:10.238: getSystemDateAndTime: Finished using Homey time-----
+0:0:4.792: Fetching (10.0.1.26) NOW image from: http://www.gwelltimes.com/manage/Upload/Pic/20147914514701.jpg
+0:0:0.147: SnapShot error (10.0.1.26): FetchError: request to http://www.gwelltimes.com/manage/Upload/Pic/20147914514701.jpg failed, reason: socket hang up
    at ClientRequest.<anonymous> (/node_modules/node-fetch/lib/index.js:1455:11)
    at ClientRequest.emit (events.js:311:20)
    at Socket.socketOnEnd (_http_client.js:460:9)
    at Socket.emit (events.js:323:22)
    at endReadableNT (_stream_readable.js:1204:12)
    at processTicksAndRejections (internal/process/task_queues.js:84:21)
+0:0:0.020: Fetch NOW error (10.0.1.26): request to http://www.gwelltimes.com/manage/Upload/Pic/20147914514701.jpg failed, reason: socket hang up
+0:0:3.138: Fetching (10.0.1.26) NOW image from: http://www.gwelltimes.com/manage/Upload/Pic/20147914514701.jpg
+0:0:0.136: SnapShot error (10.0.1.26): FetchError: request to http://www.gwelltimes.com/manage/Upload/Pic/20147914514701.jpg failed, reason: socket hang up
    at ClientRequest.<anonymous> (/node_modules/node-fetch/lib/index.js:1455:11)
    at ClientRequest.emit (events.js:311:20)
    at Socket.socketOnEnd (_http_client.js:460:9)
    at Socket.emit (events.js:323:22)
    at endReadableNT (_stream_readable.js:1204:12)
    at processTicksAndRejections (internal/process/task_queues.js:84:21)
+0:0:0.017: Fetch NOW error (10.0.1.26): request to http://www.gwelltimes.com/manage/Upload/Pic/20147914514701.jpg failed, reason: socket hang up
+0:0:1.936: getSystemDateAndTime: Finished using Homey time-----
+0:0:10.146: getSystemDateAndTime: Finished using Homey time-----
+0:0:10.130: getSystemDateAndTime: Finished using Homey time-----
+0:0:10.142: getSystemDateAndTime: Finished using Homey time-----
+0:0:10.119: getSystemDateAndTime: Finished using Homey time-----

**Comment by AdyRock** [comment]
OK, I have published another version. It seems this camera doesn't return nice failure responses so I have put the try catch around each retrieval method to hopefully catch the timeout.

**Comment by rvdblom** [comment]
Made an other log, I uninstalled the camera first and added it again.
Log ID: 2020-09-12T12:24:13.038Z
App version 1.1.4

+0:0:0.000: ====  Discovery Starting  ====
+0:0:0.174: Reply from {
  "hostname": "10.0.1.26",
  "port": "8000",
  "path": "/onvif/device_service",
  "urn": "urn:uuid:06025677-2dd2-21b2-a205-4A1599913458",
  "href": "http://10.0.1.26:8000/onvif/device_service"
}
+0:0:4.950: ====  Discovery Finished  ====
+0:0:0.018: Discovered: [
  {
    "name": "10.0.1.26",
    "data": {
      "id": "10.0.1.26",
      "port": "8000"
    },
    "settings": {
      "username": "",
      "password": "",
      "ip": "10.0.1.26",
      "port": "8000",
      "urn": "urn:uuid:06025677-2dd2-21b2-a205-4A1599913458",
      "channel": 1
    }
  }
]
+0:0:10.082: Login-----
+0:0:0.013: --------------------------
+0:0:0.012: Connect to Camera 10.0.1.26:8000 - admin
+0:0:0.195: getSystemDateAndTime: Finished using Homey time-----
+0:0:0.018: getServices: Started *****
+0:0:0.120: getServices: Finished -----
+0:0:0.015: callUpstartFunctions: Started *****
+0:0:0.426: VideoSources: [
  {
    "$": {
      "token": "VideoSourceConfigSourceToken0"
    },
    "framerate": 25,
    "resolution": {
      "width": 1920,
      "height": 1080
    }
  }
]
+0:0:0.019: ActiveSource [0] = {
  "$": {
    "token": "VideoEncoderConfigToken0"
  },
  "name": "VideoEncoderConfig0",
  "useCount": 1,
  "encoding": "H264",
  "resolution": {
    "width": 1920,
    "height": 1080
  },
  "quality": 50,
  "rateControl": {
    "frameRateLimit": 25,
    "encodingInterval": 1,
    "bitrateLimit": 1200
  },
  "multicast": {
    "address": {
      "type": "IPv4",
      "IPv4Address": "10.0.1.26"
    },
    "port": 8554,
    "TTL": 64,
    "autoStart": true
  },
  "sessionTimeout": "PT0H12M0S"
}
+0:0:0.016: VideoSource [0] = {
  "$": {
    "token": "VideoSourceConfigSourceToken0"
  },
  "framerate": 25,
  "resolution": {
    "width": 1920,
    "height": 1080
  }
}
+0:0:0.019: callUpstartFunctions: Finished -----
+0:0:0.016: CONNECTED to 10.0.1.26
+0:0:0.018: Credentials OK. Adding [
  {
    "$": {
      "token": "VideoSourceConfigSourceToken0"
    },
    "framerate": 25,
    "resolution": {
      "width": 1920,
      "height": 1080
    }
  }
]
+0:0:0.241: list_devices2: Single Sources
+0:0:3.376: Initialising CameraDevice (10.0.1.26)
+0:0:0.568: CameraDevice has been added (10.0.1.26)
+0:0:20.933: Saved Credentials
+0:0:0.002: --------------------------
+0:0:0.453: Connect to Camera 10.0.1.26:8000 - admin
+0:0:2.123: getSystemDateAndTime: Finished using Homey time-----
+0:0:0.286: getServices: Started *****
+0:0:2.813: getServices: Finished -----
+0:0:0.002: callUpstartFunctions: Started *****
+0:0:1.732: VideoSources: [
  {
    "$": {
      "token": "VideoSourceConfigSourceToken0"
    },
    "framerate": 25,
    "resolution": {
      "width": 1920,
      "height": 1080
    }
  }
]
+0:0:0.023: ActiveSource [0] = {
  "$": {
    "token": "VideoEncoderConfigToken0"
  },
  "name": "VideoEncoderConfig0",
  "useCount": 1,
  "encoding": "H264",
  "resolution": {
    "width": 1920,
    "height": 1080
  },
  "quality": 50,
  "rateControl": {
    "frameRateLimit": 25,
    "encodingInterval": 1,
    "bitrateLimit": 1200
  },
  "multicast": {
    "address": {
      "type": "IPv4",
      "IPv4Address": "10.0.1.26"
    },
    "port": 8554,
    "TTL": 64,
    "autoStart": true
  },
  "sessionTimeout": "PT0H12M0S"
}
+0:0:0.056: VideoSource [0] = {
  "$": {
    "token": "VideoSourceConfigSourceToken0"
  },
  "framerate": 25,
  "resolution": {
    "width": 1920,
    "height": 1080
  }
}
+0:0:0.065: callUpstartFunctions: Finished -----
+0:0:0.029: CONNECTED to 10.0.1.26
+0:0:0.040: getServices: Started *****
+0:0:0.317: getServices: Finished -----
+0:0:0.075: Camera Information (10.0.1.26): {
  "manufacturer": "PPS IPC based on ONVIF",
  "model": "Speed 5S",
  "firmwareVersion": "ppstrong-c51-s_neutral-2.9.5.20200205",
  "serialNumber": "060256772",
  "hardwareId": "Ver 2.1"
}
+0:0:0.027: getCapabilities: Started *****
+0:0:0.409: getCapabilities: Finished -----
+0:0:0.055: Camera (10.0.1.26) does NOT support PullPoint Events
+0:0:0.035: Supported Events(10.0.1.26): 
+0:0:0.117: Removing unsupported motion capabilities for 10.0.1.26
+0:0:1.094: Snapshot URL: http://www.gwelltimes.com/manage/Upload/Pic/20147914514701.jpg
+0:0:0.024: Registering Now image (10.0.1.26)
+0:0:0.034: SnapShot save file (10.0.1.26) = eventimage10_0_1_26.jpg
+0:0:0.048: Registering event image (10.0.1.26)
+0:0:0.025: Camera (10.0.1.26) is ready
+0:0:0.135: registered Now image (10.0.1.26)
+0:0:0.090: registered event image (10.0.1.26)
+0:0:9.319: getSystemDateAndTime: Finished using Homey time-----
+0:0:2.656: Fetching (10.0.1.26) NOW image from: http://www.gwelltimes.com/manage/Upload/Pic/20147914514701.jpg
+0:0:0.160: SnapShot error (10.0.1.26): FetchError: request to http://www.gwelltimes.com/manage/Upload/Pic/20147914514701.jpg failed, reason: socket hang up
    at ClientRequest.<anonymous> (/node_modules/node-fetch/lib/index.js:1455:11)
    at ClientRequest.emit (events.js:311:20)
    at Socket.socketOnEnd (_http_client.js:460:9)
    at Socket.emit (events.js:323:22)
    at endReadableNT (_stream_readable.js:1204:12)
    at processTicksAndRejections (internal/process/task_queues.js:84:21)
+0:0:0.026: Fetching (10.0.1.26) NOW image with Basic Auth. From: http://www.gwelltimes.com/manage/Upload/Pic/20147914514701.jpg
+0:0:0.133: SnapShot error (10.0.1.26): FetchError: request to http://www.gwelltimes.com/manage/Upload/Pic/20147914514701.jpg failed, reason: socket hang up
    at ClientRequest.<anonymous> (/node_modules/node-fetch/lib/index.js:1455:11)
    at ClientRequest.emit (events.js:311:20)
    at Socket.socketOnEnd (_http_client.js:460:9)
    at Socket.emit (events.js:323:22)
    at endReadableNT (_stream_readable.js:1204:12)
    at processTicksAndRejections (internal/process/task_queues.js:84:21)
+0:0:0.041: Fetching (10.0.1.26) NOW image with Digest Auth. From: http://www.gwelltimes.com/manage/Upload/Pic/20147914514701.jpg
+0:0:0.153: SnapShot error (10.0.1.26): FetchError: request to http://www.gwelltimes.com/manage/Upload/Pic/20147914514701.jpg failed, reason: socket hang up
    at ClientRequest.<anonymous> (/node_modules/node-fetch/lib/index.js:1455:11)
    at ClientRequest.emit (events.js:311:20)
    at Socket.socketOnEnd (_http_client.js:460:9)
    at Socket.emit (events.js:323:22)
    at endReadableNT (_stream_readable.js:1204:12)
    at processTicksAndRejections (internal/process/task_queues.js:84:21)
+0:0:0.023: Fetch NOW error (10.0.1.26): request to http://www.gwelltimes.com/manage/Upload/Pic/20147914514701.jpg failed, reason: socket hang up
+0:0:6.989: getSystemDateAndTime: Finished using Homey time-----

**Comment by AdyRock** [comment]
Hmm, I'm not sure where to go from here. I don't know of any other way to connect to grab the image as that is trying the two authentication methods that I know of. Have you setup a username and password for the camera and is there anywhere to configure external access bits?

I must admit it's the first time I have seen one with an external URL for the image though.

I checked out the support pages for the camera and it has this:

> 9. Is ONVIF supported by this camera?
> This camera does not support ONVIF.

Which is almost surprising as the spec mentions ONVIF. I say almost as it is not the first time I have come across a camera that says it is ONVIF compatible but fails. I have three different models now that are just sitting on a shelf collection dust. This one does seem to get further than most of those though.

Maybe you can contact their support and ask about ONVIF compatibility and how to fetch the image from the snapshot URL that is reported by the camera.

**Comment by rvdblom** [comment]
Thanks for your support so far!

I will send them an email with the questions you asked.

For the mean time I installed an other app on my phone and with that app I got a livestream rtsp adres

I will add the screenshots ![image](https://user-images.githubusercontent.com/42841815/93000834-29171980-f52b-11ea-93f0-85f2910281aa.png)
![image](https://user-images.githubusercontent.com/42841815/93000839-32a08180-f52b-11ea-8a5f-2b1c78ffc38e.png)

On the following screenshot is the onvif settings page from the original app
![image](https://user-images.githubusercontent.com/42841815/93000896-68456a80-f52b-11ea-968e-a57de18bc7e4.png)

When I have an answer from there support I will let you know.

Greetings, Richard
